modules/core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/computation/BulkScrollerComputation.java

```diff
@@ -272,17 +283,30 @@ protected void updateStatusAfterScroll(ComputationContext context, String comman
     /**
      * Produces a bucket as a record to appropriate bulk action stream.
      */
-    protected void produceBucket(ComputationContext context, String commandId, int bucketSize, long bucketNumber) {
+    protected void produceBucket(ComputationContext context, String commandId, int bucketSize, long bucketNumber,
+            long documentCount) {
         List<String> ids = documentIds.subList(0, min(bucketSize, documentIds.size()));
         BulkBucket bucket = new BulkBucket(commandId, ids);
         String key = commandId + ":" + Long.toString(bucketNumber);
         Record record = Record.of(key, BulkCodecs.getBucketCodec().encode(bucket));
-        if (produceImmediate) {
-            ((ComputationContextImpl) context).produceRecordImmediate(actionStream, record);
+        if (produceImmediate || (produceImmediateThreshold > 0 && documentCount > produceImmediateThreshold)) {
+            ComputationContextImpl contextImpl = (ComputationContextImpl) context;
+            if (!contextImpl.getRecords(actionStream).isEmpty()) {
+                flushRecords(contextImpl, commandId);
+            }
+            contextImpl.produceRecordImmediate(actionStream, record);
         } else {
             context.produceRecord(actionStream, record);
         }
         ids.clear(); // this clear the documentIds part that has been sent
     }
 
+    protected void flushRecords(ComputationContextImpl contextImpl, String commandId) {
+        log.warn("Scroller records threshold reached ({}) for action: {} on command: {}, flushing records downstream",
+                produceImmediateThreshold, actionStream, commandId);
+        contextImpl.getRecords(actionStream)
+                   .forEach(record -> contextImpl.produceRecordImmediate(actionStream, record));
+        contextImpl.getRecords(actionStream).clear();
```

Comment:
Can we put `contextImpl.getRecords(actionStream)` in a local variable or is the semantic wrong for that?

Reply:
we could will it be more readable not sure

Reply:
It is more readable, you're just not used to it 😀 With a local variable it's immediately clear that we're acting twice on the same object instead of re-doing a call to getRecords that may return different things... Which was a question I asked myself when I read this!

Reply:
Here, for the reasons I mentioned, the code would have been better with a local variable. Please don't ignore reviewer's comments and requests for changes.